src/libraries/System.Private.CoreLib/docs/Container Design Notes.md

```diff
@@ -0,0 +1,60 @@
+
+Passing a Dictionary of identical type to the Dictionary constructor will reuse the existing `_entries` array from the dictionary being copied if possible, so this operation also does not invoke `GetHashCode` when the source and destination comparer are identical.
+
+Documentation states that any key used in a Dictionary must not change in any way that affects its hash after insertion, so it is technically possible to remove or change this caching behavior in the future, but if a user provides a comparer with side-effects, such a change could cause breakage.
+
+### Thread-safety Violations Must Produce Exceptions
+
+Dictionary is designed to reliably produce exceptions when it is manipulated from multiple threads at once without synchronization, and we have multiple tests that verify this. "Improving" Dictionary to be able to perform operations safely across multiple threads without synchronization or throwing will as a result produce test failures and could result in observable behavior changes.
+
+It is okay for a Dictionary to become irreversibly corrupted when manipulated from multiple threads, but it should not lead to buffer overruns or other types of memory safety errors.
+
+### Specialization for Structs with Default Comparers
+
+For struct-typed keys where the default comparer is used, key loops in Dictionary are specialized to use `EqualityComparer<K>.Default` directly instead of perform a virtual invoke on a comparer instance. This is represented by having a `null` `Dictionary._comparer`. The JIT performs aggressive optimizations when this specialization is done correctly.
+
+### Prime Bucket Counts
+
+When allocating buckets, we round the number of buckets up to the next largest or equal prime number. This is done to improve collision resistance, so any change to this behavior should be carefully examined to ensure that collision resistance does not suffer.
+
+### Deterministic Serialization
+
+Serialization preserves the order of all key/value pairs in addition to the internal version of the Dictionary and the internal comparer in use.
+
+## HashSet
+
+HashSet's implementation is based on Dictionary's so it inherits most of the same guarantees and traits as that container.
+
+### Unsafe Removal During Enumeration
+
+The documentation for HashSet does *not* specify that it is safe to Remove or Clear during enumeration of a HashSet. However, the current implementation does not increment `_version` inside of Clear or Remove, so it is possible that there is code in the wild that depends on this undocumented behavior. TrimExcess *does* invalidate enumerators, as it does in Dictionary.
```

Comment:
Is this a documentation bug?

Reply:
I'm inclined to say it is and we can just fix the documentation, unless there's a reason we don't want to make this guarantee for HashSet. It seems like a useful property.